File: 09.15/프로그래머스/조수연/두 큐 합 같게 만들기.kt
package com.ssafy.algorithm

import java.util.*

fun main() {
    val sl = Solution()
    val queue1 = intArrayOf(1,1,1,1,1,1,1,1,1,10)
    val queue2 = intArrayOf(1,1,1,1,1,1,1,1,1,1)
    println(sl.solution(queue1, queue2))
}

class Solution {
    var sum1: Long = 0
    var sum2: Long = 0

    fun solution(queue1: IntArray, queue2: IntArray): Int {
        var answer: Int = -2
        var sum: Long = 0

        //큐의 합 구하기
        for (i in queue1) sum1 += i
        for (i in queue2) sum2 += i
        sum = (sum1 + sum2) / 2

        if (queue1.maxOrNull()!! > sum || queue2.maxOrNull()!! > sum) answer = -1 //인덱스가 각큐의 합보다 클 경우 -1 리턴
        else answer = shift(queue1, queue2, sum)
        return answer
    }

    fun shift(queue1: IntArray, queue2: IntArray, sum: Long): Int {

        var count = 0
        val shift1: Queue<Long> = LinkedList()
        val shift2: Queue<Long> = LinkedList()

        for (value in queue1) { //큐에 값 채우기
            shift1.offer(value.toLong())
        }

        for (value in queue2) {
            shift2.offer(value.toLong())
        }
        var time = 0
        while (time <= 300000) { //shift가 똑같이 반복하면 탈출 loop

            if (sum1 == sum2) break //합이 같으면 종료

            if (sum1 < sum) { //shift1 큐의 합이 전체 큐의 합 / 2보다 작으면 shift1에 value 추가
                val qPoll = shift2.poll()
                shift1.offer(qPoll)
                sum1 += qPoll //shift1 sum1 값 증가
                sum2 -= qPoll //shift2 sum2 값 감소
                count++
            } else {
                val qPoll = shift1.poll()
                shift2.offer(qPoll)
                sum2 += qPoll
                sum1 -= qPoll
                count++
            }
            time++
        }
        if (time > 300000){
            count = -1
        }
        return count
    }
}

File: 2022/09.15/프로그래머스/이아름/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
	int answer = 0;
	long long sum1 = 0, sum2 = 0, half = 0;
	int count = queue1.size() + queue2.size();
	queue<int> q1, q2;
	for (int i = 0; i < queue1.size(); i++) { //q에 넣기 + 총합 구하기
		sum1 += queue1[i];
		q1.push(queue1[i]);
	}
	for (int i = 0; i < queue2.size(); i++) { //q에 넣기 + 총합 구하기
		sum2 += queue2[i];
		q2.push(queue2[i]);
	}
	half = sum1 + sum2;
	if (half % 2) return -1; //홀수이면 -1
	half /= 2;
	while (answer < count * 10) { //임의의 수
		if (sum1 == half) return answer; //탈출조건
		while (sum1 > half && !q1.empty()) { //한쪽이 크면 pop
			int n = q1.front(); q1.pop();
			if (n > half) return -1;
			q2.push(n);
			sum1 -= n;
			sum2 += n;
			answer++;
		}
		while (sum2 > half && !q2.empty()) {
			int n = q2.front(); q2.pop();
			if (n > half) return -1;
			q1.push(n);
			sum2 -= n;
			sum1 += n;
			answer++;
		}
	}
	if (sum1 != half) return -1; //모두 돌아도 충족하지 않으면 -1
	return answer;
}
